nixos/modules/services/web-apps/immich.nix

```
@@ -0,0 +1,283 @@
+      };
+      port = mkOption {
+        type = types.port;
+        default = 6379;
+        description = "The port that redis will listen on.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !isUnixSocket -> cfg.secretsFile != null;
+        message = "A secrets file containing at least the database password must be provided when unix sockets are not used.";
+      }
+    ];
+
+    services.postgresql = mkIf cfg.database.enable {
+      enable = true;
+      ensureDatabases = mkIf cfg.database.createDB [ cfg.database.name ];
+      ensureUsers = mkIf cfg.database.createDB [
+        {
+          name = cfg.database.user;
+          ensureDBOwnership = true;
+          ensureClauses.login = true;
+        }
+      ];
+      extraPlugins = mkIf cfg.database.setupPgvectors (ps: with ps; [ pgvecto-rs ]);
+      settings = mkIf cfg.database.setupPgvectors {
+        shared_preload_libraries = [ "vectors.so" ];
+        search_path = "\"$user\", public, vectors";
+      };
+    };
+    systemd.services.postgresql.serviceConfig.ExecStartPost =
+      let
+        sqlFile = pkgs.writeText "immich-pgvectors-setup.sql" ''
+          CREATE EXTENSION IF NOT EXISTS vectors;
+          CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
+          ALTER SCHEMA vectors OWNER TO ${cfg.database.user};
+          GRANT SELECT ON TABLE pg_vector_index_stat TO ${cfg.database.user};
+
+          ALTER EXTENSION vectors UPDATE;
+        '';
+      in
+      mkIf cfg.database.setupPgvectors [
+        ''
+          ${config.services.postgresql.package}/bin/psql -d "${cfg.user}" -f "${sqlFile}"
+        ''
+      ];
+
+    services.redis.servers = mkIf cfg.redis.enable {
+      "${cfg.redis.serverName}" = {
+        enable = true;
+        port = cfg.redis.port;
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.port ];
+
+    systemd.services.immich-server = {
+      description = "Immich backend server (Self-hosted photo and video backup solution)";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = lib.mkMerge [
+        postgresEnv
+        {
+          REDIS_HOSTNAME = cfg.redis.host;
+          HOST = cfg.host;
+          IMMICH_PORT = toString cfg.port;
+          IMMICH_MEDIA_LOCATION = cfg.mediaLocation;
+          IMMICH_MACHINE_LEARNING_URL = "http://localhost:3003";
+        }
+        cfg.environment
+      ];
+
+      serviceConfig = commonServiceConfig // {
+        ExecStart = lib.getExe cfg.package;
+        EnvironmentFile = mkIf (cfg.secretsFile != null) cfg.secretsFile;
+        StateDirectory = "immich";
+        User = cfg.user;
+        Group = cfg.group;
+      };
+    };
+
+    systemd.services.immich-machine-learning = mkIf cfg.machine-learning.enable {
+      description = "immich machine learning";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
```

```suggestion
      wantedBy = [ "multi-user.target" ];
      environment = cfg.machine-learning.environment;
```
The option is otherwise not used.